validator/gulpjs/index.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * @license
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the license.
+ */
+
+'use strict';
+
+const through = require('through2');
+const gutil = require('gulp-util');
+const amphtmlValidator = require('amphtml-validator');
+
+const PLUGIN_NAME = 'gulp-amphtml-validator';
+const PluginError = gutil.PluginError;
+
+/**
+ * Validates AMP files and attaches the validation result to the file object.
+ *
+ * @param {Object} validator - amphtml validator
+ * @returns {stream} gulp file stream
+ */
+module.exports.validate = function(validator) {
+
+  if (!validator) {
+    validator = amphtmlValidator;
+  }
+
+  function runValidation(file, encoding, callback) {
+    if (file.isNull()) {
+      return callback(null, file);
+    }
+    if (file.isStream()) {
+      this.emit('error', new PluginError(PLUGIN_NAME,
+        'Streams not supported!'));
+    }
+    if (file.isBuffer()) {
+      const context = this;
+      validator.getInstance()
+        .then(function(validatorInstance) {
+          const inputString = file.contents.toString();
+          file.ampValidationResult = validatorInstance.validateString(inputString);
+          return callback(null, file);
+        })
+        .done(function(result) {
+          if (result instanceof Error) {
+            context.emit('error', new PluginError(PLUGIN_NAME,
```

Comment:
So this is the validator itself failing, e.g., if it can't load validator.js or similar. To which extent is this distinguished from files that fail validation?
Would it be useful to document this distinction at the top level of the API? I think it could help people who use this the first time and have some problem with their setup / don't know yet how it looks when it's working. I think it's perfectly fine if you solve this problem in whatever way is natural for gulpjs, I just want to make sure that the three cases "something wrong with validator" / "file validates" / "file doesn't validate" are somewhat understood by a user. I understand that "something wrong with validator" may be rare.